### PR TITLE
feature: Neural Network Batching

### PR DESCRIPTION
**Pull request by Zephreo**
Added mini-batches to NN library.

Averages the gradients across a batch.

Currently slower than no batching, but is non-dependent on each other during a loop and therefore can be processed in parallel as a performance improvement later.

MicrosoftTeams-image

### CODE REVIEW

**Comment by Zephreo** [comment]
Requesting review since I take the average of target subtract the average of the batch? Does that bias the target to be average and therefore make loss irrelevent?

**Comment by Zephreo** [comment]
After splitting train process performance has increased.
These results are consistent for me.
image

**Review by Alex-MobileApps — APPROVED** [review]
This looks good Harry. The average loss and gradients are being calculated correctly for mini-batch updates, and the update is correctly applied only at the end of the batch. All other test cases are working as expected
